### PR TITLE
Week 3

### PR DESCRIPTION
**Pull request by darishenko**
Add dynamic Attribute; Implement custom Events and Listeners; Create CronJobs; Define Validation Constraints; Localize Attribute values and UI; Add Media Files; Define Default Property Values

### CODE REVIEW

File: bin/custom/concerttours/resources/concerttours-items.xml [hunk 1]
```diff
@@ -117,6 +122,10 @@
                     <description>type of concert (indoors or open air)</description>
                     <persistence type="property"/>
                 </attribute>
+                <attribute qualifier="daysUntil" type="java.lang.Long">
```
Comment (devjohnsmith1) [inline bin/custom/concerttours/resources/concerttours-items.xml:125]: naming of an attribute is not clear: days until what? Isn't it better to rename to smth like daysUntillConcertLeft? + including a handler

File: bin/custom/concerttours/resources/concerttours-items.xml [hunk 2]
```diff
@@ -140,5 +149,35 @@
...
+                    <description>short headline for the news item</description>
+                    <persistence type="property"/>
+                </attribute>
+                <attribute qualifier="content" type="java.lang.String">
+                    <description>fuller description of the news item</description>
+                    <persistence type="property"/>
+                </attribute>
+            </attributes>
+        </itemtype>
+        <itemtype code="NotLoremIpsumConstraint" extends="AttributeConstraint"
```
Comment (devjohnsmith1) [inline bin/custom/concerttours/resources/concerttours-items.xml:169]: why do we need a Constraint that checks for the exact text? Is it a check for some default text? Unclear

File: bin/custom/concerttours/src/concerttours/events/AlbumSalesEventListener.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package concerttours.events;
+
+import java.util.Date;
+import concerttours.model.NewsModel;
+import de.hybris.platform.servicelayer.event.impl.AbstractEventListener;
+import de.hybris.platform.servicelayer.model.ModelService;
+import reactor.util.annotation.NonNull;
```
Comment (devjohnsmith1) [inline bin/custom/concerttours/src/concerttours/events/AlbumSalesEventListener.java]: use javax.annotation.NonNull instead

Comment (devjohnsmith1) [reply bin/custom/concerttours/src/concerttours/events/AlbumSalesEventListener.java]: + check for other occurances in code

File: bin/custom/concerttours/src/concerttours/facades/impl/DefaultBandFacade.java [hunk 4]
```diff
@@ -9,20 +9,26 @@
...
+import de.hybris.platform.servicelayer.config.ConfigurationService;
+import de.hybris.platform.servicelayer.media.MediaService;
 import org.springframework.beans.factory.annotation.Required;
 import reactor.util.annotation.NonNull;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Objects;
-import java.util.Optional;
+import java.util.*;
```
Comment (devjohnsmith1) [inline bin/custom/concerttours/src/concerttours/facades/impl/DefaultBandFacade.java]: .* imports all the classes of a package, which is not required here, please avoid this. You can check the IDEA settings for imports

File: bin/custom/concerttours/src/concerttours/facades/impl/DefaultBandFacade.java [hunk 5]
```diff
@@ -124,4 +146,14 @@ private List<AlbumData> getBandAlbumData(Long bandPk) {
         return albumDataList;
     }
 
+    protected String getImageURL(final BandModel bandModel, final MediaFormatModel format) {
```
Comment (devjohnsmith1) [inline bin/custom/concerttours/src/concerttours/facades/impl/DefaultBandFacade.java:161]: @Nullable

File: bin/custom/concerttours/web/webroot/WEB-INF/views/bandDetails.jsp [hunk 6]
```diff
@@ -5,6 +5,9 @@
 <body>
 <h1>Band Details</h1>
```
Comment (devjohnsmith1) [inline bin/custom/concerttours/web/webroot/WEB-INF/views/bandDetails.jsp]: please, put all string values in jsp's under localization properties and load on frontend with spring:message

**Review by devjohnsmith1 — CHANGES_REQUESTED** [review]

File: bin/custom/concerttours/resources/impex/essentialdata-cronJobs.impex
Comment (darishenko) [inline bin/custom/concerttours/resources/impex/essentialdata-cronJobs.impex:1]: 0/30 min

File: bin/custom/concerttours/resources/impex/essentialdata-cronJobs.impex [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# Define the cron job and the job that it wraps
+INSERT_UPDATE CronJob; code[unique=true];job(code);singleExecutable;sessionLanguage(isocode)
+;sendNewsCronJob;sendNewsJob;false;de
+
+# Define the trigger that periodically invokes the cron job
+INSERT_UPDATE Trigger;cronjob(code)[unique=true];cronExpression
+#% afterEach: impex.getLastImportedItem().setActivationTime(new Date());
+; sendNewsCronJob; 0/10 * * * * ?
```
Comment (darishenko) [inline bin/custom/concerttours/resources/impex/essentialdata-cronJobs.impex]: 0/30 min

**Review by devjohnsmith1 — APPROVED** [review]

**Pull request merged**